---
layout: post
title:  "Learn Python 3 the Hard Way: Exercise 16 - で学んだこと"
---

[Learn Python 3 the Hard Way](https://learnpythonthehardway.org/)の「Exercise 16 - 」で学んだことのまとめです。

## Exercise 16: Reading and Writing Files

```python
# * 覚えるべきコマンド
# close - ファイルを保存して閉じる。
# read - ファイルの内容を返す
# readline - ファイルの内容を１行だけ返す(=改行がある箇所まで) 
# truncate - ファイルを空にする
# write(’stuff’) - ファイルに"stuff"を書き込む 
# seek(0)- 読み込み/書き込みし始める箇所をファイルの先頭に移動

# touchでTest_Sexy.txtファイルを作ってから
# ❯ python3.6 ex16/ex16.py Test_Sexy.txtを実行

from sys import argv

script, filename, = argv

# We're going to erase Test_Sexy.txt.
# If you don't want that, hit CTRL-C (^C)
# If you do want that, hit RETURN.
print(f"We're going to erase {filename}.") 
print("If you don't want that, hit CTRL-C (^C)")
print("If you do want that, hit RETURN.")

# ?
input("?")
# →terminalでreturnキー

# * ここでctrl+cを押した場合は
# ?^CTraceback (most recent call last):
#  File "ex16/ex16.py", line 17, in <module>
#    input("?")
# KeyboardInterrupt

# Opening the file...
# Truncating the file. Goodbye!
print("Opening the file...")
target = open(filename)
print("Truncating the file. Goodbye!")
target.truncate()
# * openの中にパラメーター'w'がないと
# Traceback (most recent call last):
#   File "ex16/ex16.py", line 38, in <module>
#     target.truncate()
# io.UnsupportedOperation: File not open for writing
# =ファイルがwritingのために開かれてないよ！とエラーになる


#Now, I'm going to ask you for three lines.
print("Now, I'm going to ask you for three lines.")

line1 = input("line 1: ")
line2 = input("line 2: ")
line3 = input("line 3: ")
# terminalで
# line 1: 佐藤勝利は
# line 2: Sexy Zoneの
# line 3: プロ意識担当
# を入力

# I'm going to write these to the file.
print("I'm going to write these to the file.")

target.write(line1)
target.write("\n")
target.write(line2)
target.write("\n")
target.write(line3)
target.write("\n")

#And finally, we close it.
print("And finally, we close it.")
target.close()
```
↓書き込まれたファイル"Test_Sext.txt"

![](https://cl.ly/1S2g2p3V3b1T/[c882727198db2d13c5daf065976e5bb3]_Image%202017-10-04%20at%209.44.16%20PM.png)

## Exercise 17: More Files

```python17
# ❯ python3.6 ex17/ex17.py test_sexy.txt test_zone.txt
# で実行

from sys import argv
from os.path import exists


script, from_file, to_file = argv

# Copying from test_sexy.txt to test_zone.txt
print(f"Copying from {from_file} to {to_file}")

# we could do rhese two on one line, how?
in_file = open(from_file)
indata = in_file.read()
# * readすることでtxtファイルの内容を変数にする

# The input file is 24 bytes long
print(f"The input file is {len(indata)} bytes long")
# * len() - 文字数(改行も1文字に含める)

# Does the output file exist? True
print(f"Does the output file exist? {exists(to_file)}")
# path が実在する場合 True を返す

# Ready, hit RETURN to continue, control-C to abort.
print("Ready, hit RETURN to continue, control-C to abort.")
input()

out_file = open(to_file, 'w')
out_file.write(indata)

#Alright, all done.
print("Alright, all done.")

out_file.close()
in_file.close()
```

```python
# * ↑の動作をシンプルに実行する方法

from sys import argv
script, in_file_name, out_file_name = argv

out_file = open(out_file_name, 'w')
in_file = open(in_file_name)

out_file.write(in_file.read())

# openした時の変数が無いとcloseできない！！
out_file.close()
in_file.close()

# * built in function...
#   前に.がつかない(open(), len(), input(), print(), ...)
# * method...
#   前に.がつく(.write(), .close(), .read(),)
#   .の前には変数がないといけない
#   つまり .close(out_file) はだめ


# * open(x)は変数xにファイル名を使う
# * y.write(...)の変数yは、openでできた変数出なければいけない
# * y.read(), y.close()も同じ
```

## Exercise 18: More 

```python

```
